lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,405 @@
+#ifndef IRODS_TICKET_API_HPP
+#define IRODS_TICKET_API_HPP
+
+#include <string_view>
+#include <string>
+
+#undef NAMESPACE_IMPL
+#undef RxComm
+#undef rxTicketAdmin
+
+#ifdef IRODS_RESOURCE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#    define NAMESPACE_IMPL server
+#    define RxComm         RsComm
+#    define rxTicketAdmin  rsTicketAdmin
+
+struct RsComm;
+#else
+#    define NAMESPACE_IMPL client
+#    define RxComm         RcComm
+#    define rxTicketAdmin  rcTicketAdmin
+
+struct RcComm;
+#endif // IRODS_RESOURCE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+
+enum ticket_type
+{
+    READ,
+    WRITE
+};
+
+// WIP: user_constraint, host_constraint, and group_constraint
+enum ticket_prop
+{
+    USES,
+    WRITE_FILE,
+    WRITE_BYTE
+};
+enum ticket_operation
+{
+    SET,
+    REMOVE
+};
+
+class constraint
```

Comment:
Here is an example of how you may implement constraint types. Again, this is just one way to implement this. There are other ways to achieve this, but I'll leave that up to you for investigation.

The idea here is to leverage the type system and overload `add_ticket_constraint` and `remove_ticket_constraint`.

The use of `explicit` on the constructors is important here. You can read about that here: https://en.cppreference.com/w/cpp/language/explicit
```cpp
#include <string_view>

struct RxComm {};

struct user_constraint
{
    explicit user_constraint(std::string_view _v)
        : value{_v}
    {
    }

    std::string_view value;
};

struct group_constraint
{
    explicit group_constraint(std::string_view _v)
        : value{_v}
    {
    }
    
    std::string_view value;
};

struct host_constraint
{
    explicit host_constraint(std::string_view _v)
        : value{_v}
    {
    }
    
    std::string_view value;
};

auto add_ticket_constraint(RxComm& _comm, const user_constraint& _constraint) -> void
{
    // Apply user constraint.
}

auto add_ticket_constraint(RxComm& _comm, const group_constraint& _constraint) -> void
{
    // Apply group constraint.
}

auto add_ticket_constraint(RxComm& _comm, const host_constraint& _constraint) -> void
{
    // Apply host constraint.
}

auto remove_ticket_constraint(RxComm& _comm, const user_constraint& _constraint) -> void
{
    // Remove user constraint.
}

auto remove_ticket_constraint(RxComm& _comm, const group_constraint& _constraint) -> void
{
    // Remove group constraint.
}

auto remove_ticket_constraint(RxComm& _comm, const host_constraint& _constraint) -> void
{
    // Remove host constraint.
}

int main()
{
    RxComm comm;

    add_ticket_constraint(comm, user_constraint{"rods"});
    add_ticket_constraint(comm, group_constraint{"special_group"});
    add_ticket_constraint(comm, host_constraint{"example.irods.org"});

    remove_ticket_constraint(comm, group_constraint{"special_group"});
}
```